test/slim_check.lean

```diff
@@ -0,0 +1,69 @@
+
+import system.random.basic
+import data.nat.prime
+import data.zmod.basic
+
+/-- fermat's primality test -/
+def primality_test (p : ℕ) (h : fact (0 < p)) : rand bool :=
+if h : 2 ≤ p-1 then do
+  n ← rand.random_r 2 (p-1) h,
+  return $ (n : zmod p)^(p-1) = 1 -- we do arithmetic with `zmod n` so that modulo and multiplication are interleaved
+else return (p = 2)
+
+/-- `iterated_primality_test_aux p h n` generating `n` candidate witnesses that `p` is a
+composite number and concludes that `p` is prime if none of them is a valid witness  -/
+def iterated_primality_test_aux (p : ℕ) (h : fact (0 < p)) : ℕ → rand bool
+| 0 := pure tt
+| (n+1) := do
+  b ← primality_test p h,
+  if b
+    then iterated_primality_test_aux n
+    else pure ff
+
+def iterated_primality_test (p : ℕ) : rand bool :=
+if h : 0 < p
+  then iterated_primality_test_aux p h 10
+  else pure ff
+
+/-- `find_prime_aux p h n` generates `n` candidate prime numbers,
+tests each and 20 of their successors with `iterated_primality_test`
+and return the first one that can pass the primality test. -/
```

Comment:
I'm having trouble parsing this doc-string. What does `each and 20 of their successors` mean?

Reply:
The approach is to pick one random odd number, test if it's prime, if it fails, add 2 and try again. You do that 20 times and then you give up on this random number. You pick a new one, test it, test the 19 odd numbers after that one and then generate a new random number.